db/models/projectComponent.js

```diff
@@ -1,30 +1,152 @@
-ProjectComponentModel.getByProject = (projectId) => {};
-ProjectComponentModel.getByUser = (userId) => {};
+ProjectComponent.create = ({userId, projectId, type, resourceUrl, name, description, isDownloadable = false}) => {
+  if (!componentTypes[type]) {
+    return Promise.reject('Component type is invalid');
+  }
+  return User.getById(userId)
+    .then(() => {
+      return Project.getById(projectId);
+    })
+    .then((project) => {
+      return Project.getContributors(projectId)
+        .then((contributors) => {
+          let isContributor = !!contributors.filter((user) => {
+            return user.id === userId;
+          })[0];
+          if (!isContributor && project.ownerId !== userId) {
+            return Promise.reject('Cannot create a component in a project you are not a part of');
+          }
+        });
+    })
+    .then(() => {
+      return ProjectComponent.model.create({
+        authorId: userId,
+        projectId,
+        resourceUrl,
+        name,
+        description,
+        type,
+        isDownloadable: isDownloadable === true ? true : false
+      });
+    });
+};
+
+ProjectComponent.update = (userId, componentId, options) => {
+  return User.getById(userId)
+    .then(() => {
+      return ProjectComponent.getById(componentId);
+    })
+    .then((component) => {
+      return Project.getById(component.projectId)
+        .then((project) => {
+          return Project.getContributors(project.id)
+          .then((contributors) => {
+            let isContributor = !!contributors.filter((user) => {
+              return user.id === userId;
+            })[0];
+            if (!isContributor && project.ownerId !== userId) {
+              return Promise.reject('Cannot edit a component in a project you are not a part of');
+            }
+            return component.update(options);
+          });
+        });
+      });
+};
+
+ProjectComponent.delete = (userId, componentId) => {
+  return User.getById(userId)
+    .then(() => {
+      return ProjectComponent.getById(componentId);
+    })
+    .then((component) => {
+      return Project.getById(component.projectId)
+        .then((project) => {
+          return Project.getContributors(project.id)
+          .then((contributors) => {
+            let isContributor = !!contributors.filter((user) => {
+              return user.id === userId;
+            })[0];
+            if (!isContributor && project.ownerId !== userId) {
+              return Promise.reject('Cannot delete a component in a project you are not a part of');
+            }
+            return component.destroy();
+          });
+        });
+      })
+      .then(() => {
+        return true;
+      });
+};
+ProjectComponent.getByProject = (projectId) => {
+  return Project.getById(projectId)
+    .then(() => {
+      return ProjectComponent.model.findAll({
+        where: {projectId}
+      });
+    });
+};
+
+// TODO - Implement this method
```

Comment:
Please do not leave old commented out code. Basically your codebase is the final version of your book that you want to send out to the world. You don't want little notes in the margins and highlights all over the place.  This commented out code makes it more difficult to read through this code base and creates appearance of this code still being in development, not yet ready for prod.